array.c

```diff
@@ -4692,198 +4692,251 @@ rb_ary_clear(VALUE ary)
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, -2, 1) # => ["a", "b", :X, "d"]
+ *  - If +start+ is too small (less than and far from zero), replaces all elements:
  *
- *  If +start+ is large (<tt>start >= array.size</tt>), extends +self+ with +nil+:
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', -9)     # => ["-", "-", "-", "-"]
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, 5, 0) # => ["a", "b", "c", "d", nil]
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, 5, 2) # => ["a", "b", "c", "d", nil, :X, :X]
+ *  With argument +object+ and numeric arguments +start+ and +length+,
+ *  replaces elements based on the given +start+ and +length+:
  *
- *  If +length+ is zero or negative, replaces no elements:
+ *  - If +start+ is in range, replaces +length+ elements beginning at offset +start+:
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, 1, 0) # => ["a", "b", "c", "d"]
- *    a.fill(:X, 1, -1) # => ["a", "b", "c", "d"]
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', 1, 2)   # => ["a", "-", "-", "d"]
  *
- *  With arguments +obj+ and Range +range+, and no block given,
- *  replaces elements based on the given range.
+ *  - If +start+ is negative, counts from the end:
  *
- *  If the range is positive and ascending (<tt>0 < range.begin <= range.end</tt>),
- *  replaces elements from <tt>range.begin</tt> to <tt>range.end</tt>:
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', -3, 2)  # => ["a", "-", "-", "d"]
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (1..1)) # => ["a", :X, "c", "d"]
+ *  - If +start+ is large (<tt>start >= array.size</tt>), extends +self+ with +nil+:
  *
- *  If <tt>range.first</tt> is negative, replaces no elements:
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', 6, 2)   # => ["a", "b", "c", "d", nil, nil, "-", "-"]
+ *      # Extends even if length is zero.
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', 6, 0)   # => ["a", "b", "c", "d", nil, nil]
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (-1..1)) # => ["a", "b", "c", "d"]
+ *  - If +length+ is zero or negative, replaces no elements:
  *
- *  If <tt>range.last</tt> is negative, counts from the end:
+ *      a = ('a'..'d').to_a # => ["a", "b", "c", "d"]
+ *      a.fill('-', 2, 0)   # => ["a", "b", "c", "d"]
+ *      a.fill('-', 2, -1)  # => ["a", "b", "c", "d"]
  *
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (0..-2)) # => [:X, :X, :X, "d"]
- *    a = ['a', 'b', 'c', 'd']
- *    a.fill(:X, (1..-2)) # => ["a", :X, :X, "d"]
+ *  With argument +object+ and Range argument +range+,
+ *  replaces elements based on the given range:
  *
- *  If <tt>range.last</tt> and <tt>range.last</tt> are both negative,
- *  both count from the end of the array:
+ *  - If the range is positive and ascending (<tt>0 < range.begin <= range.end</tt>),
+ *    replaces elements from <tt>range.begin</tt> to <tt>range.end</tt>:
```

Comment:
This is not always true, it depends on if the end is inclusive or exclusive, for example:

```ruby
["a", "b", "c", "d"].fill("-", 1..2) # => ["a", "-", "-", "d"]
["a", "b", "c", "d"].fill("-", 1...2) # => ["a", "-", "c", "d"]
(1..2).begin # => 1
(1..2).end # => 2
(1...2).begin # => 1
(1...2).end # => 2
```

Reply:
Agreed;  needs revision.